src/app/components/utility/userStateService.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * Created by chima on 11/19/15.
+ */
+'use strict'
+
+angular.module('utility')
+  .service('userStateService', function (ehaCouchDbAuthService, locationService, $localForage, $q) {
+    var self = this
+    self.stateMap = {states: [], selectedState: ''}
+
+    function initializeStateVariables () {
+      $localForage.clear()
+      self.stateMap.states = []
+      self.stateMap.selectedState = ''
+    }
+
+    this.getUserSelectedState = function (byId) {
+      return ehaCouchDbAuthService.getCurrentUser()
+        .then(function (user) {
+          if (!user.userCtx.name) {
+            return ''
+          }
+          return $localForage.getItem(user.userCtx.name)
+            .then(function (state) {
+              var result = !byId ? state : (byId === true ? self.getState(state, true) : self.getState(state))
+              return $q.when(result)
+            })
+        })
+    }
+
+    this.setUserSelectedState = function (state) {
+      return ehaCouchDbAuthService.getCurrentUser()
+        .then(function (user) {
+          if (!state || !user.userCtx.name) {
+            return $q.when('')
+          }
+          return $localForage.setItem(user.userCtx.name, state)
+            .then(function (st) {
+              return true
+            })
+        })
+    }
+
+    this.getState = function (name, byId) {
```

Comment:
Remember, when a user refreshes the page, the states variable resets to undefined even if the user is already logged in. And we cant be fetching the states when the user visits the site for the first time since we don't even know who the user is and the states he has access to at that point. what i did is fetch the user states at login time and cache it using localForage till the user logs out.